Lato Sensu (2025) - I.A Generativa: Conceitos e Aplicações


 Uso de n-gramas em NLP



In [ ]:

# Instala bibliotecas
!pip install nltk==3.5

import nltk
nltk.download("punkt")

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [ ]:
# Passo 1: Carregar o corpus

texto = """Na praça havia um banco
Havia um banco na praça
Um banco estava na praça
Na praça estava um banco
Na praça havia um banco tranquilo
O banco na praça era verde
Havia um banco perto da fonte
Um banco descansava sob a árvore
Na praça havia bancos espalhados
O banco estava vazio na praça
Bancos verdes e tranquilos na praça
Havia um banco com flores ao lado"""

# Transformando em lista de linhas
texto = texto.lower().split("\n")
print("Corpus tokenizado por linhas:")
print(texto)
print("Número de versos:", len(texto))

Corpus tokenizado por linhas:
['na praça havia um banco', 'havia um banco na praça', 'um banco estava na praça', 'na praça estava um banco', 'na praça havia um banco tranquilo', 'o banco na praça era verde', 'havia um banco perto da fonte', 'um banco descansava sob a árvore', 'na praça havia bancos espalhados', 'o banco estava vazio na praça', 'bancos verdes e tranquilos na praça', 'havia um banco com flores ao lado']
Número de versos: 12


In [ ]:
# Passo 2: Tokenizando as sentenças

lista_tokens = []
for verso in texto:
    tokens = nltk.word_tokenize(verso, language="portuguese")
    lista_tokens.append(tokens)

print("\nLista de tokens por verso:")
print(lista_tokens)
print("Número de versos tokenizados:", len(lista_tokens))


Lista de tokens por verso:
[['na', 'praça', 'havia', 'um', 'banco'], ['havia', 'um', 'banco', 'na', 'praça'], ['um', 'banco', 'estava', 'na', 'praça'], ['na', 'praça', 'estava', 'um', 'banco'], ['na', 'praça', 'havia', 'um', 'banco', 'tranquilo'], ['o', 'banco', 'na', 'praça', 'era', 'verde'], ['havia', 'um', 'banco', 'perto', 'da', 'fonte'], ['um', 'banco', 'descansava', 'sob', 'a', 'árvore'], ['na', 'praça', 'havia', 'bancos', 'espalhados'], ['o', 'banco', 'estava', 'vazio', 'na', 'praça'], ['bancos', 'verdes', 'e', 'tranquilos', 'na', 'praça'], ['havia', 'um', 'banco', 'com', 'flores', 'ao', 'lado']]
Número de versos tokenizados: 12


In [ ]:
#  Pré-processamento e n-gramas

from nltk.lm.preprocessing import padded_everygram_pipeline
from nltk.lm import Laplace

ngramas = 2  # bigramas
# Cria pipeline de n-gramas e vocabulário
train_data, vocab = padded_everygram_pipeline(ngramas, lista_tokens)

In [ ]:
# Treina o  modelo  com Laplace smoothing

lm = Laplace(ngramas)
lm.fit(train_data, vocab)


In [ ]:
# Testando o modelo

# Gerar texto a partir de uma palavra semente
print("\nExemplo de geração de 6 tokens a partir de 'um':")
print(lm.generate(6, text_seed=["um"]))

# Probabilidade da palavra "banco" dado o contexto "um"
score = lm.score("banco", context=["um"])
logscore = lm.logscore("banco", context=["um"])
print(f"\nProbabilidade de 'banco' dado o contexto ['um']: {score}")
print(f"Log-probabilidade de 'banco' dado o contexto ['um']: {logscore}")




Exemplo de geração de 6 tokens a partir de 'um':
['banco', 'descansava', 'sob', 'a', 'árvore', '</s>']

Probabilidade de 'banco' dado o contexto ['um']: 0.23684210526315788
Log-probabilidade de 'banco' dado o contexto ['um']: -2.0780025120012735


##Instala e importa bibliotecas

Instala o NLTK, que é uma biblioteca que ajuda a trabalhar com texto.

Baixa um pacote chamado punkt, que serve para separar o texto em palavras.


##Carrega o texto (corpus)
Cada linha do texto é tratada como uma frase separada.

O texto é colocado todo em minúsculas.


##Tokeniza as frases
Cada frase é dividida em palavras.

Exemplo: "Na praça havia um banco" vira ['na', 'praça', 'havia', 'um', 'banco'].


##Pré-processamento e criação de bigramas

Cria pares de palavras (bigramas), como ('um', 'banco').

Adiciona sinais de início <s> e fim </s> de frase.

Define todas as palavras que o modelo vai conhecer (vocabulário).


##Treina o modelo de linguagem

Usa o modelo Laplace, que é um tipo de modelo que evita que a probabilidade de algo seja zero, mesmo se nunca apareceu antes.

O modelo aprende quais palavras aparecem depois de outras palavras no corpus.


##Testa o modelo

Gera uma sequência de 6 palavras começando com "um".

Calcula a probabilidade de a palavra "banco" aparecer depois de "um".

Mostra também a log-probabilidade, que é só a mesma ideia da probabilidade, mas em escala logarítmica.

Ex:
####Probabilidade: 0.2368
####Log-probabilidade: -2.078

Isso quer dizer que cerca de 24% das vezes, "banco" vem depois de "um" no texto que o modelo aprendeu.

Log-probabilidade é só outra forma de mostrar a mesma informação.

Quanto mais próximo de 1 a probabilidade, mais provável a palavra aparecer.